### ISSUE TITLE
Using Mujoco Contact Model/Collision Detection to Validate Geom Attachement

### ISSUE THREAD

**Issue by GerritBartels** [deepmind/mujoco#881]
Hi,

I'm a student and I'm trying to use MuJoCo for a large university project. I am writing an algorithm to generate ```.xml``` files. As randomness is involved in creating these files (e.g. sampling x,y coordinates for placing a cube), I need to validate that the geoms are not colliding with each other/keep a minimum distance. As this seems far from trivial to implement from scratch, especially with the plan of also validating placement of meshes, I am trying to use Mujoco's internal contact model/collision detection to validate the placement of geoms. One rather imperfect workflow I have come up with is the following:

Say I have an mjcf object called ```mujoco_object``` (e.g. a cube) that I loaded from this ```.xml``` file:

```xml
<mujoco model="cube">
    <worldbody>
        <body name="cube" pos="0 1 1">
            <geom type="box" size="1 1 1" rgba="1. 0 0 1"/>
        </body>
    </worldbody>
</mujoco>
```

as well as an mjcf object called ```site``` (i.e. the environment that I later want to save as ```.xml``` file) loaded from this ```.xml``` file:

```xml
<mujoco model="site">
    <default>
        <default class="/"/>
        <default class="cube/"/>
    </default>
    <worldbody>
        <body name="cube/">
            <body name="cube/cube" pos="0 0 1">
                <geom name="cube//unnamed_geom_0" class="cube/" type="box" size="1 1 1" rgba="1 0 0 1"/>
            </body>
        </body>
    </worldbody>
</mujoco>
```

I can then execute the following code to check whether ```mujoco_object``` has any contacts with other geoms in ```site``` by adding a joint and setting its .margin property (I found that the joint has to be added as the ```ncon``` function would not return any contacts otherwise).

```python
from dm_control import mjcf

def validate(mujoco_object, site, minimum_distance=0.1):
    mujoco_object.worldbody.body[0].add("joint")
    mujoco_object.worldbody.body[0].geom[0].margin = minimum_distance

    site.attach(mujoco_objectj)

    physics = mjcf.Physics.from_mjcf_model(site)
    num_contacts = physics.data.ncon

    mujoco_object.detach()

    if num_contacts == 0:
        return True
    else:
        return False

mujoco_object = mjcf.from_file("cube.xml")
site = mjcf.from_file("site.xml")

validate(mujoco_object, site)
```
 --> For this example the function would evaluate to ```False```

This, however, is not only quite hacky but also computationally expensive as I have to add a joint to the object, attach it to the site, create a physics instance, check for contacts, and then detach it to later attach the original ```mujoco_object``` with valid position and without the added joint. 

Is there a better way to do this? I have looked a lot at the Mujoco docs and the mujoco/dm_control source code, but I have not found an alternative for doing this. Any help would be greatly appreciated!

**Comment by yuvaltassa** [comment]
The reason you need a joint is that otherwise both geoms are static and therefore counted as "in the same body" for collision purposes (the world body, in this case).

Basically, you are doing the right thing.  The only easy alternative I can think of is to let the physics do the job for you. Put all you bodies in the scene with free joints (or just 3 linear sliders), turn off gravity and introduce high fluid viscosity. Simulate for some time until all the contacts are resolved, then save the new positions. This could be after, depending on the number of objects for which you need to call your `validate` function.

The other option would be to call MuJoCo's low-level collision functions, as mentioned in [this reply](https://github.com/deepmind/mujoco/issues/872#issuecomment-1556175328) to a very similar request. Unfortunately the low level collision functions are not currently exposed to Python, but I hope this will happen in a forthcoming release (cc @saran-t). Of course if you switch to C/C++ you will be able to call these functions, but I guess that's not an option.
